File: src/main/java/net/cakemc/database/cursor/DefaultCursor.java
package net.cakemc.database.cursor;

import net.cakemc.database.api.Piece;
import net.cakemc.database.filter.Filter;

import java.util.Comparator;
import java.util.List;

/**
 * The type Default cursor.
 */
public class DefaultCursor extends Cursor<Piece> {

    private List<Piece> pieces;

    /**
     * Instantiates a new Default cursor.
     *
     * @param pieces the pieces
     */
    public DefaultCursor(List<Piece> pieces) {
        this.pieces = pieces;
    }

    /**
     * Gets pieces.
     *
     * @return the pieces
     */
    public List<Piece> getPieces() {
        return pieces;
    }


    @Override
    public Cursor<Piece> limit(int number) {
        this.pieces = pieces.stream().limit(number).toList();
        return this;
    }

    @Override
    public Cursor<Piece> sort(Comparator<Piece> comparator) {
        pieces.sort(comparator);
        return this;
    }

    @Override
    public int index(Piece element) {
        return pieces.indexOf(element);
    }

    @Override
    public int index(Filter<Piece> filter) {
        Piece single = pieces.stream()
                .filter(filter::matches)
                .findFirst()
                .orElse(null);

        if (single == null)
            return -1;

        return pieces.indexOf(single);
    }

    @Override
    public List<Piece> collect() {
        return pieces;
    }
}

File: src/main/kotlin/net/cakemc/database/cursor/DefaultCursor.kt
package net.cakemc.database.cursor

import net.cakemc.database.api.Piece
import net.cakemc.database.filter.Filter

/**
 * The type Default cursor.
 */
class DefaultCursor
/**
 * Instantiates a new Default cursor.
 *
 * @param pieces the pieces
 */(
    /**
     * Gets pieces.
     *
     * @return the pieces
     */
    var pieces: List<Piece?>
) : Cursor<Piece?>() {

    override fun limit(number: Int): Cursor<Piece?> {
        this.pieces = pieces.stream().limit(number.toLong()).toList()
        return this
    }

    override fun sort(comparator: Comparator<Piece?>): Cursor<Piece?> {
        pieces.sortedWith(comparator)
        return this
    }

    override fun index(element: Piece?): Int {
        return pieces.indexOf(element)
    }

    override fun index(filter: Filter<Piece?>): Int {
        val single = pieces.stream()
            .filter { document: Piece? -> filter.matches(document) }
            .findFirst()
            .orElse(null)

        if (single == null) return -1

        return pieces.indexOf(single)
    }

    override fun collect(): List<Piece?> {
        return pieces
    }
}
